### DISCUSSION TITLE
Load SENEC storage when price is low via HA automations

### DISCUSSION THREAD

**Discussion by marq24** [marq24/ha-senec-v3#113] [Show and tell]
![previw](https://github.com/marq24/ha-senec-v3/raw/main/images/storage_board.png)
https://github.com/marq24/ha-senec-v3/blob/master/docs/battery_control/BATTERY_WITH_TIBBER.md

**Comment by KneterPeter** [comment]
![Screenshot_20241215-120355.png](https://github.com/user-attachments/assets/94f0bb7a-f7e5-42cc-9f13-954f4b814d12)

Senec hat da auch was neues. Kam bei mir heute.
Hoffentlich kommt das auch für die "älteren"
Senec v3 hybrid duo

**Reply by marq24** [reply]
... naja - also ob ich den Implementierungskünsten von Senec wirklich Vetrauen möchte...

**Reply by mschmiedel** [reply]
Am Ende ist ein HA Skript, welches die Wallbox auf "Schnell Laden" bei niedrigem Strompreis einstellt, kein Hexenwerk. Interessant wird es ja erst, wenn es offiziell erlaubt ist, den Speicher aus dem Netz vollzuladen...

**Reply by marq24** [reply]
@mschmiedel "von welcher Mama" möchtest Du denn die Erlaubnis haben Deinen Speicher mit Strom aus dem Netz zu laden? Mein Kenntnisstand ist, dass man aus dem Speicher nicht ins Netz einspeisen darf... (was meist finanziell sowieso wenig Sinn ergibt)...

Und es geht hier nicht um die Steuerung der Wallbox... sondern eben um die Steuerung des SENEC Heimspeichers...

**Reply by mschmiedel** [reply]
Naja, laut MaStR ist der Speicher zur Ladung aus PV Strom und Teileinspeisung berechtigt, ansonsten kann man Volleinspeisung wählen. Andere Möglichkeiten eigentlich nicht wählbar. Das ist im EEG schon eine Grauzone, denn theoretisch darf man ja keinen nicht-EE Strom einspeisen. Es wäre natürlich besser, wenn der Gesetzgeber die Speicher ganz aus dieser Rechnung nehmen würde. Der Senec Speicher erlaubt theoretisch keine Einspeisung, aber regelt meist so auf -20W bis -30W, sodass kein Strom aus dem Netz kommt. Rein theoretisch kann ja so um die 0,7kWh pro Tag eingespeist werden.

Ob das ggf. beim Netzbetreiber auffallen kann oder nicht, ist eine andere Geschichte.

**Comment by UDicke** [comment]
Hallo Matthias,
vielen Dank für das Thema und die ausführliche Beschreibung [docs/battery_control/BATTERY_WITH_TIBBER.md](https://github.com/marq24/ha-senec-v3/blob/master/docs/battery_control/BATTERY_WITH_TIBBER.md).
Ich denke damit würde ich sehr gut zurecht kommen in Verbindung mit einem Ampere.StoragePro Speicher.

2 Punkte habe ich:
- ich sehe keine "Entladesperre" für Zeiten wo der Preis etwas über dem aktuellen MinPreis liegt. Ich halte so eine Sperre für wichtig um nicht zu niedrigem Preis zu laden und sofort wieder zu entladen. Das habe ich bei mir über recht einfache Automatisierungen gelöst. 
  sensor.Preis_niedrig wird "True" wenn aktueller Preis <= 1,1 * aktueller MinPreis ist. Sensor.Preis_niedrig schaltet die Entladesperre in HA.
  Oder hab ich was übersehen?
- ich würde lieber nur dann die Batterie laden wenn der Preis min oder nahe min ist. Spricht da was gegen? Meine Idee:
  Bis der Optimizer in evcc funktioniert und auch meine Wünsche erfüllt, würde ich gerne einen Loadpoint und Ladeplan für die Hausbatterie (als Benutzerdefiniertes Gerät) anlegen. und das Ergebnis zur Steuerung nutzen. Im Ladeplan kann ich z.B. einen MaxSoc konfigurieren und bekomme die notwendigen (günstigsten) Ladefenster abhängig vom Ladestand der Batterie geliefert. Im Moment habe ich aber noch keine Idee wie man Daten zur Steuerung auslesen/nutzen kann. Ob ich evtl. Daten aus [marq24/ha-evcc](https://github.com/marq24/ha-evcc) nutzen kann habe ich noch nicht gefunden.
<img width="1113" height="585" alt="grafik" src="https://github.com/user-attachments/assets/9e84f40f-f812-44ae-989a-ab1d30203678" />
<img width="972" height="601" alt="grafik" src="https://github.com/user-attachments/assets/4db4375c-b790-42ad-8779-2baea714e32d" />

Ich bin gespann auf Deine Kommentare...

**Reply by marq24** [reply]
Wäre cool wenn der Senec Speicher so was hätte wie eine Entladesperre - deswegen habe ich das nicht umgesetzt - das iat aber tatsächlich ein essentielles feature... 

Und "ja" über die evcc HA integration solltest Du eigentlich auch alles machen können (und den Ladeplan kannst Du über Services modifizieren)....

**Reply by UDicke** [reply]
> Wäre cool wenn der Senec Speicher so was hätte wie eine Entladesperre - deswegen habe ich das nicht umgesetzt - das iat aber tatsächlich ein essentielles feature...

Mein Speicher hat auch keine explizite Entladesperre.
Der läuft normal im "Backup-Mode" und es gibt ein Register für den min. Backup Soc. Da setzte ich den Wert hoch. So funktioniert auch die Entladesperre für die Hausbatterie innerhalb von evcc.
Alternativ könnte ich dem Register für den min SOC der Batterie einen hohen Wert geben. Das funktioniert dann auch im sogenannten "Self-Use Mode".
Ich habe in der evcc Doku gesehen, dass die aktive Batteriesteuerung für Senec da ist, allerdings ohne Entladesperre.
Bei anderen Herstellern wird die Entladesperre über das Setzen des Entladestroms für die Batterie auf 0 Ampere oder KW gesetzt. Kann der Senec denn keine von diesen Optionen? Geht denn das nicht über die "SPARE_CAPACITY"?
> Und "ja" über die evcc HA integration solltest Du eigentlich auch alles machen können (und den Ladeplan kannst Du über Services modifizieren)....

Ein echter Ladeplan für die Hausbatterie existiert ja in evcc (noch) nicht.  Im Moment nutze ich die Möglichkeit number.evcc_battery_grid_charge_limit zu schreiben um das Laden der Hausbatterie innerhalb evcc zu schalten. Ich suche jetzt noch die Möglichkeit die Daten vom "Plan" zu nutzen um damit zu steuern. Da habe ich noch keine Idee.

**Comment by ngaertner** [comment]
Hallo @marq24,

ich habe die Tibber Ladelogik per HA Automation bei mir jetzt auch umgesetzt, nachdem unser V2.1 mit 5KW letzte Woche endlich auf einen V3 mit 10KW LFP getauscht wurde - vorher hat sich das nicht wirklich gelohnt.

Was mir nach zwei Tagen Test auffällt ist, dass er anscheinend bei der Berechnung des Endes einer "Cheapest" Periode immer eine Stunde daneben liegt. Bilde ich mir das nur ein, oder ist da irgendwo noch ein Bug?

<img width="1017" height="612" alt="image" src="https://github.com/user-attachments/assets/962021a0-d8fc-4b40-9aec-62215e916ace" />

Ansonsten ist deine Logik wirklich großartig! Hilft mir auch sehr viel meine zwei SENEC Wallboxen nachts effizient nur in den Billigsten Zeiträumen zu aktivieren.

**Reply by marq24** [reply]
Hi Nico,

also natürlich kann in diesen ganzen Scripten noch ein Bug sein - gerade nach der Umstellung auf den 15min Takt... 

Du hast recht, das zwar der niedrigste Preis irgendwo um die 5 Uhr sitzt... Aber das "Billigpreis-Fenster" hört ja nicht am tiefsten Preis auf, sondern es geht natürlich solange weiter, wie der Preis unter dem zur Berechnung zugrunde liegenden Durchschnittspreis liegt... 

Allerdings muss ich hier auch nochmal genauer gucken - bei mir gibt es noch eine Hochpreis & Niedrigpreis Fenster, ein "Billigpreis" habe ich aktuell bei mir in der GUI gar nicht (mehr?)

**Reply by ngaertner** [reply]
Hi Matthias,

den "Billigpreis" habe ich selbst eingebaut - sind aber dieselben Werte die bei "Volle Ladung" angezeigt werden (ist mir erst hinterher aufgefallen).

"solange weiter, wie der Preis unter dem zur Berechnung zugrunde liegenden Durchschnittspreis liegt":
Die Magie dahinter ist mir noch nicht ganz klar - ich habe mir den Code angesehen, ist aber mein erstes Jinja template und daher schwer zu entziffern.

Ich war jetzt pauschal davon ausgegangen, dass er das Keywort "CHEAP" in den Tibber Daten mit berücksichtigt.
Zumindest beim Anfang des Fensters (01:15) scheint es eine Rolle zu spielen (oder ist das Zufall?), aber beim Ende scheint er (relativ konsequent nach meinen Beobachtungen, kann aber auch Zufall sein), drüber raus zu schießen.

```
- total: 0.2868
  startsAt: '2025-12-02T01:00:00.000+01:00'
  level: NORMAL
- total: 0.2785
  startsAt: '2025-12-02T**01:15:00**.000+01:00'
  level: CHEAP
- total: 0.2763
  startsAt: '2025-12-02T01:30:00.000+01:00'
  level: CHEAP
...
- total: 0.2688
  startsAt: '2025-12-02T05:00:00.000+01:00'
  level: CHEAP
- total: 0.2795
  startsAt: '2025-12-02T05:15:00.000+01:00'
  level: CHEAP
- total: 0.2913
  startsAt: '2025-12-02T05:30:00.000+01:00'
  level: NORMAL
- total: 0.3036
  startsAt: '2025-12-02T05:45:00.000+01:00'
  level: NORMAL
- total: 0.2834
  startsAt: '2025-12-02T06:00:00.000+01:00'
  level: NORMAL
- total: 0.3002
```

Viele Grüße,

Nico

**Reply by ngaertner** [reply]
Noch ein Kommentar: wenn die Logik wirklich mit einem Durchschnittspreis arbeitet, dann ist da definitiv noch ein kleines Problem in der Logik (siehe die Werte oben).

Der Durchschnittspreis müsste in dem Beispiel ja über 0.2785 (01:15 = start) und unter 0.2868 (01:10) gelegen haben, aber er hat am Ende noch die Werte 0.2913 (05:30) und 0.3036 (05:45 = ende) in die Phase mit einbezogen, obwohl die beiden Werte über 0.2868 liegen.

**Reply by marq24** [reply]
> "solange weiter, wie der Preis unter dem zur Berechnung zugrunde liegenden Durchschnittspreis liegt": Die Magie dahinter ist mir noch nicht ganz klar - ich habe mir den Code angesehen, ist aber mein erstes Jinja template und daher schwer zu entziffern.

Das ist echt ganz schön 'ekliger' Code diese template Soße... 

Ich müsste jetzt nochmal ganz tief in meinen alten grauen Zellen graben, um das korrekt zusammen zu bringen... aber das Thema "was ist der aktuelle Durchschnittspreis" ist nicht trivial...

Deswegen rechne ich (meine ich) mit unterschiedlichen Durchschnitten - Es gibt zwar den von Tibber berechneten 24h Durchschnittspreis, aber dieser Wert ist meiner Meinung nach nicht hilfreich... Denn das ist nicht der Durchschnittspreis (D) der nächsten 24h - sondern der D. des aktuellen Tages von 0:00 bis 23:59... 

Ich meine mein D. mit dem ich den kram berechne guckt immer nur in die Zukunft. Aber das habe ich auch min. 4 mal geändert.

- tibb_avg_price (Tages Durchschnitt vom Tibber Sensor)
- average_price (Heute+Morgen Durchschnitt)
- avgprice_ahead (Heute+Morgen Durchschnitt - nur ab "jetzt")

Während ich da jetzt so durch meinen Code wühle, verstehe ich mehr oder minder NIX mehr - der Tag war wohl einfach zu lang - sorry...


> Ich war jetzt pauschal davon ausgegangen, dass er das Keywort "CHEAP" in den Tibber Daten mit berücksichtigt. Zumindest beim Anfang des Fensters (01:15) scheint es eine Rolle zu spielen (oder ist das Zufall?), aber beim Ende scheint er (relativ konsequent nach meinen Beobachtungen, kann aber auch Zufall sein), drüber raus zu schießen.

Diese keywords wie CHEAP verwende ich gar nicht... [ich meine ich wollte das selbst bestimmen]

**Reply by ngaertner** [reply]
Macht nichts - bisher funktioniert es eigentlich ganz Ok, und wegen den paar ct ein Fass aufzumachen lohnt sich nicht.

Vielleicht noch ein paar Anmerkungen:

- In deiner Action zum einschalten verwendest du "states('sensor.pow_storage_full_charge_duration')|int > 0.49"
--> das kann meiner Meinung nach nicht richtig funktionieren, da 0.49 ja kein "int" ist - ich habs bei mir jetzt mal durch "states('sensor.pow_storage_full_charge_duration')|round(2) > 0.49" ersetzt (auch wenn mir die Logik dahinter nicht ganz klar ist)
- Ich habe meine Einschalt Action generell so geändert, dass er nur dann das Netzladen startet wenn der Speicher schon unter 65% entladen ist ( ich spiele aktuell noch mit dem Wert)  und wenn der Akku aktuell nicht (bzw. nicht viel) durch Solarüberschuss beladen wird - damit verhindere ich, dass die Akkubeladung nicht tagsüber los läuft solange noch genug Sonne runterkommt um den Hausbedarf zu decken, und der Akkustand prinzipiell noch durch die Sonne steigen kann.

```
{{ 
  states('switch.senec_safe_charge')|lower == 'off' and
  states('binary_sensor.pow_pv_power_can_fully_load_storage')|lower == 'off' and
  (
    ( states('sensor.senec_battery_charge_percent')|int <= 65 and
      states('sensor.senec_battery_charge_power')|int < 50  and
      states('binary_sensor.pow_now_in_cheapest_phase')|lower == 'on'
    )
    and
    (      
      states('sensor.pow_storage_full_charge_duration')|round(2) > 0.49 and
      (
        ( states('binary_sensor.pow_now_in_cheapest_phase')|lower == 'on')
        or
        ( states('sensor.pow_storage_full_charge_duration') >= states('sensor.pow_low_price_duration') ) 
      )
    )
  )	
}}
```
- Das Abschalten kommt bei mir bei einem Ladestand >= 80% (das reicht bei mir normal locker um über den Tag zu kommen, und lässt noch genug Luft um den spärlichen Solarüberschuss im Winter einzusammeln). Außerdem schalte ich schon ab wenn "pow_now_in_low_price_phase" false ist, und nicht nur wenn der Preis im "High Price" ist.

```
{{ 
  states('switch.senec_safe_charge')|lower == 'on' and (
  states('sensor.senec_battery_charge_percent')|int >= 80 or (
  states('binary_sensor.pow_now_in_high_price_phase')|lower == 'on' or
  states('binary_sensor.pow_now_in_low_price_phase')|lower == 'off' or  
  states('binary_sensor.pow_now_in_high_price_storage_independent_phase')|lower == 'on' or
  states('binary_sensor.pow_pv_power_can_fully_load_storage')|lower =='on' ) )
}}
```
Im Grunde genommen ist die ganze Berechnung und Steuerung ein eigenes Projekt.
Wenn's mich irgendwann mal in den Fingern juckt werde ich mich selbst mal dransetzen und das aufschreiben.

**Reply by ngaertner** [reply]
PS: Ich verwende deine POW Sensoren auch um meine Wallboxen nachts zur Niedrigpreiszeit ein- und wieder auszuschalten - funktioniert aktuell super!

--> ein (Strom muss billig und in "cheapest" phase sein)
```
{{ 
  states('sensor.tibber_strompreis')|round(1) < 31 and
  states('binary_sensor.pow_now_in_cheapest_phase')|lower == 'on'
}}
```

--> aus (Strom darf nicht teuer sein oder in "cheapest" phase sein)
```
{{ 
  states('sensor.tibber_strompreis')|round(1) > 32 or
  states('binary_sensor.pow_now_in_cheapest_phase')|lower == 'off'
}}
```

**Reply by dgeffe66** [reply]
Moin

Ich versuche das in gang zu bekommen.
Scheider aber an den Sensoren.
Ich habe jetzt folgende
            sensor.zu_hause_preis_nachste_12h
          - sensor.zu_hause_aktueller_strompreis
          - sensor.senec_battery_charge_percent
          - sensor.solcast_pv_forecast_prognose_verbleibende_leistung_heute 
         Aber ich glaube hier einen Fehler zu haben.
